src/librustc_span/hygiene.rs

```diff
@@ -1170,13 +1170,27 @@ pub fn raw_encode_expn_id<E: Encoder>(
     mode: ExpnDataEncodeMode,
     e: &mut E,
 ) -> Result<(), E::Error> {
-    if !context.serialized_expns.lock().contains(&expn) {
-        context.latest_expns.lock().insert(expn);
-    }
+    // Record the fact that we need to serialize the corresponding
+    // `ExpnData`
+    let needs_data = || {
+        if !context.serialized_expns.lock().contains(&expn) {
+            context.latest_expns.lock().insert(expn);
+        }
+    };
+
     match mode {
-        ExpnDataEncodeMode::IncrComp => expn.0.encode(e),
+        ExpnDataEncodeMode::IncrComp => {
+            // Always serialize the `ExpnData` in incr comp mode
+            needs_data();
+            expn.0.encode(e)
+        }
         ExpnDataEncodeMode::Metadata => {
             let data = expn.expn_data();
+            // We only need to serialize the ExpnData
+            // if it comes from this crate.
+            if data.krate == LOCAL_CRATE {
```

Comment:
What about the case of proc macro dependencies, for which span metadata is inlined?

Reply:
We currently serialize all `SyntaxContexts` as `SyntaxContext::root` for proc macro crates, since they're not really observable:

https://github.com/rust-lang/rust/blob/8e738539be23a62120059b5b4443f6c235f932b4/src/librustc_metadata/rmeta/encoder.rs#L255-L286

I'll add a comment noting this